posts/2020-02-12-using-dotty-union-types-with-akka-typed-part-II.adoc

```diff
@@ -0,0 +1,296 @@
+
+NOTE: One could say that from the outside, the internal behavior is not utilised to its full extent.
+
+The combination of Dotty Union types combined with the generic `Behavior` of having a contravariant type parameter leads to a very simple implementation of the `Pinger` actor. The future will tell if, with Dotty, Akka will be able to exploit this in a future version.
+
+== Returning to best practices for coding up actors in Akka Typed
+
+The coding style I'm using is drawn from the https://doc.akka.io/docs/akka/current/typed/style-guide.html#style-guide[Akka Typed Style guide]. This guide leaves some choices to the programmer such as choosing between an _object oriented style_ or a _functional style_. For the functional style, I prefer to put the core behavior of a typed actor in a companion class. An advantage of this approach is that the method that will return the initial behavior doesn't have to take extra contextual parameters that need to be passed in: these can be added as class parameters. In simple cases, that may be considered overkill, but as a counter argument, I think that applying the same practice in a consistent manner helps to maintain a recurring and easily recognisable way of coding actors.
+
+The Dotty version of the `Pinger` actor will then look as follows:
+
+[source,scala,linenums]
+----
+import akka.actor.typed.{ActorRef, Behavior}
+import akka.actor.typed.scaladsl.{ActorContext, Behaviors}
+
+object Pinger {
+
+  // My protocol
+  sealed trait Command
+  case object SendPing extends Command
+  case object StopPingPong extends Command
+
+  // My protocol + the responses I need to understand...
+  private type CommandsAndResponses = Command | PingPong.Response
+
+  def apply(pingPong: ActorRef[PingPong.Ping]): Behavior[Command] = {
+    val internalBehavior = Behaviors.setup[CommandsAndResponses] { context =>
+      (new Pinger(context, pingPong)).run()
+    }
+    internalBehavior.narrow
+  }
+}
+
+class Pinger private (context: ActorContext[Pinger.CommandsAndResponses], pingPong: ActorRef[PingPong.Ping]) {
+
+  import Pinger._
+
+  def run(): Behavior[CommandsAndResponses] =
+    Behaviors.receiveMessage {
+      case StopPingPong =>
+        context.log.info(s"End of the ping-pong game - calling it a day!")
+        context.system.terminate()
+        Behaviors.stopped
+      case SendPing =>
+        pingPong ! PingPong.Ping(replyTo = context.self)
+        Behaviors.same
+      case response : PingPong.Response =>
+        context.log.info(s"Hey: I just received a $response !!!")
+        Behaviors.same
+    }
+}
+----
+
+Note that we prevent the direct creation of instances of the `Pinger` actor by marking the constructor of the `Pinger` class private.
+
+=== Conclusions
+
+In this article we have:
+
+* Shown in both Scala 2 and Dotty that we can hide the handling of responses to messages sent to other actors:
```

Comment:
In this article, you've shown how to handle responses sent to other actors without unnecessarily widening the message protocol.